File: python/triangle/triangle.py
from itertools import permutations


def is_triangle(sides: list[int]) -> bool:
    """
    Checks for the Triangle Inequality Theorem to
    determine if the sides form a triangle.
    """
    for a, b, c in permutations(sides):
        if a + b <= c:
            return False

    return True


def equilateral(sides: list[int]) -> bool:
    if not is_triangle(sides):
        return False
    return len(set(sides)) == 1


def isosceles(sides: list[int]) -> bool:
    if not is_triangle(sides):
        return False
    return len(set(sides)) <= 2


def scalene(sides: list[int]) -> bool:
    if not is_triangle(sides):
        return False
    return len(set(sides)) == 3

File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn is-valid? [side-a side-b side-c]
  (and (> (+ side-a side-b) side-c)
       (> (+ side-a side-c) side-b)
       (> (+ side-b side-c) side-a)))

(defn equilateral? [side-a side-b side-c]
  (and (is-valid? side-a side-b side-c) (== side-a side-b side-c)))

(defn isosceles? [side-a side-b side-c]
  (and (is-valid? side-a side-b side-c)
       (or (== side-a side-b)
       (== side-a side-c)
       (== side-b side-c))))
  

(defn scalene? [side-a side-b side-c]
  (and (is-valid? side-a side-b side-c)
       (not (equilateral? side-a side-b side-c))
       (not (isosceles? side-a side-b side-c))))
